x-pack/filebeat/input/awss3/sqs_acker.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package awss3
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common/acker"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+)
+
+// EventACKTracker tracks the publishing state of S3 objects. Specifically
+// it tracks the number of message acknowledgements that are pending from the
+// output. It can be used to wait until all ACKs have been received for one or
+// more S3 objects.
+type EventACKTracker struct {
+	DeletionWg *sync.WaitGroup
+
+	EventsToBeAcked  *atomic.Uint64
+	TotalEventsAcked *atomic.Uint64
+
+	ackMutex             *sync.RWMutex
+	ackMutexLockedOnInit *atomic.Bool
+	isSQSAcker           bool
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	msg             *types.Message
+	ReceiveCount    int
+	start           time.Time
+	processingErr   error
+	Handles         []s3ObjectHandler
+	keepaliveCancel context.CancelFunc
+	keepaliveWg     *sync.WaitGroup
+	msgHandler      sqsProcessor
+	log             *logp.Logger
+	client          beat.Client
+}
+
+func NewEventACKTracker(ctx context.Context, deletionWg *sync.WaitGroup) *EventACKTracker {
+	ctx, cancel := context.WithCancel(ctx)
+	ackMutex := new(sync.RWMutex)
+	// We need to lock on ack mutex, in order to know that we have passed the info to the acker about the total to be acked
+	// Lock it as soon as we create the acker. It will be unlocked and in either SyncEventsToBeAcked or AddSQSDeletionData
+	ackMutex.Lock()
+	return &EventACKTracker{
+		ctx:                  ctx,
+		cancel:               cancel,
+		ackMutex:             ackMutex,
+		ackMutexLockedOnInit: atomic.NewBool(true),
+		DeletionWg:           deletionWg,
+		TotalEventsAcked:     atomic.NewUint64(0),
+		EventsToBeAcked:      atomic.NewUint64(0),
+	}
+}
+
+func (a *EventACKTracker) SyncEventsToBeAcked(s3EventsCreatedTotal uint64) {
+	// We want to execute the logic of this call only once, when the ack mutex was locked on init
+	if !a.ackMutexLockedOnInit.Load() {
```

Comment:
I'm unclear how `ackMutexLockedOnInit` is supposed to work:
- It's checked and reset both here and in `AddSQSDeletionData` below. Are these two functions supposed to be mutually exclusive? If so, could there be comments clarifying that (maybe including where the flag is defined)?
- It's an atomic, but it isn't modified atomically. Can these functions be called more than once concurrently? If not, then there's no need for the flag to be atomic. If so, then the check at the beginning of the function should be: `if !a.ackMutexLockedOnInit.Swap(false) ...`

Reply:
> I'm unclear how `ackMutexLockedOnInit` is supposed to work:
> 
> * It's checked and reset both here and in `AddSQSDeletionData` below. Are these two functions supposed to be mutually exclusive? If so, could there be comments clarifying that (maybe including where the flag is defined)?

yes, they are mutually exclusive: one is used when the tracker is in the sqs-s3 mode (`AddSQSDeletionData`),  and one for the s3 directly listing mode (`SyncEventsToBeAcked`): i will rename them in order to make it clear and add the comments you asked :)


> * It's an atomic, but it isn't modified atomically. Can these functions be called more than once concurrently? If not, then there's no need for the flag to be atomic. If so, then the check at the beginning of the function should be: `if !a.ackMutexLockedOnInit.Swap(false) ...`

yes, the functions can be called concurrently more than once, thanks of the head up on using `Swap` 👍